app/models/generic_event.rb

```diff
@@ -143,14 +143,43 @@ def self.details_attributes(*attributes)
     end
   end
 
+  # NB: Majority of events will use this serializer rather than the anonymous class.
+  def self.serializer
+    GenericEventSerializer
+  end
+
   # Relationship attributes live against the details but are expected in the json:api relationship section
   # so are defined separately
   def self.relationship_attributes(attributes)
     define_singleton_method(:relationship_attributes) do
       instance_variable_get('@relationship_attributes')
     end
+
     instance_variable_set('@relationship_attributes', attributes)
 
+    define_singleton_method(:serializer) do
+      @serializer ||=
+        begin
+          serializer = Class.new(GenericEventSerializer)
+
+          relationship_attributes.each do |attribute_key, attribute_type|
+            # NB: Default to V2 relationship serializer and fallback to unversioned serializer. Never support v1
+            relation_serializer = "V2::#{attribute_type.to_s.singularize.camelize}Serializer"
+            relation_serializer = if const_defined?(relation_serializer)
+                                    relation_serializer
+                                  else
+                                    relation_serializer.gsub('V2::', '')
+                                  end
+
+            named_relationship_key = attribute_key.to_s.sub('_id', '')
```

Comment:
There's quite a bit of stuff going on here that might be worth adding coverage at unit level - would be good to see specs for V1/V2 serializers being returned depending on which things we have defined.

Reply:
Re testing: I am exercising every serializer implicitly in a successful request for every event. I could add an explicit test of the serializer results for one that implements a v2 and one that doesn't have a v2?

I could also dump the result of every event serialization process to a fixture file for each event and verify that the result is what we expect?

Reply:
A simple unit level spec for this model that tests V2 fallback behaviour is all that's missing I think - happy with the higher level coverage of each generate serializer. The unit level spec also acts as living documentation so I think is beneficial to explicitly cover this behaviour.